File: frontend/src/api/documents.js
import apiClient from './index'

// 문서 관리 API
const documentsApi = {
  /**
   * 문서 목록 조회
   * @param {Object} params - 조회 파라미터
   */
  list(params = {}) {
    return apiClient.get('/api/admin/v1/documents', { params })
  },

  /**
   * 문서 상세 조회
   * @param {number} docId - 문서 ID
   */
  get(docId) {
    return apiClient.get(`/api/admin/v1/documents/${docId}`)
  },

  /**
   * 문서 저장 (임베딩 없이)
   * @param {Object} data - 문서 데이터
   */
  save(data) {
    const payload = {
      title: data.title,
      doc_type: data.docType,
      content: data.content,
      language: data.language || 'ko',
      metadata: data.metadata || {},
      context_data: data.contextData || null,
      source_type: data.sourceType || 'ui_input',
      source_file: data.sourceFile || null,
      usage_type: data.usageType || 'rag_knowledge'
    }
    if (data.tenantId) payload.tenant_id = data.tenantId
    return apiClient.post('/api/admin/v1/documents', payload)
  },

  /**
   * 문서 수정
   * @param {number} docId - 문서 ID
   * @param {Object} data - 수정할 데이터
   */
  update(docId, data) {
    const updateData = {}
    if (data.title !== undefined) updateData.title = data.title
    if (data.docType !== undefined) updateData.doc_type = data.docType
    if (data.content !== undefined) updateData.content = data.content
    if (data.language !== undefined) updateData.language = data.language
    if (data.metadata !== undefined) updateData.metadata = data.metadata
    if (data.contextData !== undefined) updateData.context_data = data.contextData

    return apiClient.put(`/api/admin/v1/documents/${docId}`, updateData)
  },

  /**
   * 문서 삭제
   * @param {number} docId - 문서 ID
   */
  delete(docId) {
    return apiClient.delete(`/api/admin/v1/documents/${docId}`)
  },

  /**
   * 문서 일괄 삭제
   * @param {number[]} docIds - 문서 ID 목록
   */
  bulkDelete(docIds) {
    return apiClient.post('/api/admin/v1/documents/bulk-delete', {
      doc_ids: docIds
    })
  },

  /**
   * 임베딩 실행
   * @param {Object} params - 임베딩 파라미터
   */
  executeEmbedding(params) {
    const payload = {
      doc_ids: params.doc_ids,
      delete_original: params.delete_original || false
    }
    if (params.chunk_size) payload.chunk_size = params.chunk_size
    if (params.chunk_overlap) payload.chunk_overlap = params.chunk_overlap
    return apiClient.post('/api/admin/v1/documents/embedding/execute', payload)
  },

  /**
   * 파일 업로드 (PDF/DOCX → 텍스트 추출)
   * @param {File} file - 업로드할 파일 객체
   * @param {Function} onProgress - 업로드 진행률 콜백 (optional)
   */
  upload(file, onProgress) {
    const formData = new FormData()
    formData.append('file', file)
    return apiClient.post('/api/admin/v1/documents/upload', formData, {
      headers: { 'Content-Type': undefined },
      ...(onProgress && {
        onUploadProgress: (e) => onProgress(Math.round((e.loaded * 100) / e.total))
      })
    })
  },

  /**
   * 청킹 미리보기
   * @param {Object} params - 미리보기 파라미터
   */
  previewChunks(params) {
    const payload = { content: params.content }
    if (params.chunk_size) payload.chunk_size = params.chunk_size
    if (params.chunk_overlap) payload.chunk_overlap = params.chunk_overlap
    return apiClient.post('/api/admin/v1/documents/embedding/preview', payload)
  }
}

export default documentsApi

File: app/models/documents.py
"""
문서 관리 스키마

문서 CRUD, 청킹, 임베딩 관련 모든 모델
Inner class를 사용하여 관련 스키마를 그룹화
"""
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# ===================================
# 문서 생성/저장
# ===================================

class DocumentCreate(BaseModel):
    """문서 생성 (기본)"""
    title: str
    doc_type: str
    language: str = "ko"
    content: str
    metadata: Optional[Dict[str, Any]] = Field(default_factory=dict)


class DocumentCreateWithChunking(BaseModel):
    """문서 생성 (자동 청킹 지원)"""
    title: str = Field(..., min_length=1, max_length=500, description="문서 제목")
    doc_type: str = Field(..., description="문서 유형 (policy, guide, faq, notice)")
    content: str = Field(..., min_length=1, description="문서 내용")
    language: str = Field(default="ko", description="언어 (ko, en)")
    metadata: Optional[Dict[str, Any]] = Field(default=None, description="메타데이터")
    auto_chunk: bool = Field(default=True, description="자동 청킹 여부")
    chunk_size: int = Field(default=1000, ge=100, le=5000, description="청크 크기 (문자 수)")
    chunk_overlap: int = Field(default=100, ge=0, le=500, description="청크 간 중복 (문자 수)")

    model_config = {
        "json_schema_extra": {
            "example": {
                "title": "2024년 보안 규정",
                "doc_type": "policy",
                "content": "제1조 목적... (긴 텍스트를 입력하세요)",
                "language": "ko",
                "metadata": {"year": 2024, "department": "Security"},
                "auto_chunk": True,
                "chunk_size": 1000,
                "chunk_overlap": 100
            }
        }
    }


class DocumentCreateResponse(BaseModel):
    """문서 생성 응답 (청킹 지원)"""
    success: bool
    message: str
    parent_id: int
    chunk_ids: List[int]
    total_chunks: int
    total_chars: int


class DocumentSaveRequest(BaseModel):
    """문서 저장 요청 (임베딩 없이 저장)"""
    title: str = Field(..., min_length=1, max_length=500, description="문서 제목")
    doc_type: str = Field(..., description="문서 유형 (policy, guide, faq, notice)")
    content: str = Field(..., min_length=1, description="문서 내용")
    language: str = Field(default="ko", description="언어 (ko, en)")
    metadata: Optional[Dict[str, Any]] = Field(default=None, description="메타데이터")
    context_data: Optional[str] = Field(default=None, description="임베딩 제외 컨텍스트 데이터 (SQL, 스키마 등 Agent 참조용)")
    source_type: str = Field(default="ui_input", description="소스 타입 (ui_input, pdf, web, api)")
    source_file: Optional[str] = Field(default=None, description="원본 파일명")
    usage_type: str = Field(default="rag_knowledge", description="문서 용도 (rag_knowledge, rag_action)")
    tenant_id: Optional[str] = Field(default=None, description="소속 테넌트 ID (GLOBAL: 선택, TENANT: 자동 부여)")

    model_config = {
        "json_schema_extra": {
            "example": {
                "title": "2024년 운영 가이드",
                "doc_type": "guide",
                "content": "본 가이드는 시스템 운영 절차를 설명합니다...",
                "language": "ko",
                "metadata": {"year": 2024, "department": "IT"},
                "source_type": "ui_input"
            }
        }
    }


class DocumentSaveResponse(BaseModel):
    """문서 저장 응답"""
    success: bool
    message: str
    doc_id: int
    title: str
    content_length: int
    needs_chunking: bool
    recommended_chunks: int


# ===================================
# 문서 조회/응답
# ===================================

class DocumentResponse(BaseModel):
    """문서 상세 응답"""
    id: int
    title: str
    doc_type: str
    language: str
    content: str
    metadata: Dict[str, Any]
    indexed: bool
    created_at: datetime
    updated_at: datetime


class DocumentListItem(BaseModel):
    """문서 목록 항목"""
    id: int
    title: str
    doc_type: str
    usage_type: str = "rag_knowledge"  # 문서 용도 (rag_knowledge, rag_action)
    language: str
    content_length: int
    original_length: Optional[int] = None  # 원본 문서 전체 길이 (청킹된 경우 원본 길이)
    source_type: Optional[str] = None
    source_file: Optional[str] = None
    total_chunks: Optional[int] = None
    indexed: bool = False
    embedded_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    tenant_id: Optional[str] = None  # 소속 테넌트 ID ('1'=공용)


class DocumentListResponse(BaseModel):
    """문서 목록 응답"""
    total: int
    items: List[DocumentListItem]


# ===================================
# 문서 수정/삭제
# ===================================

class DocumentUpdateRequest(BaseModel):
    """문서 수정 요청"""
    title: Optional[str] = Field(None, min_length=1, max_length=500, description="문서 제목")
    doc_type: Optional[str] = Field(None, description="문서 유형")
    content: Optional[str] = Field(None, min_length=1, description="문서 내용")
    language: Optional[str] = Field(None, description="언어 (ko, en)")
    metadata: Optional[Dict[str, Any]] = Field(None, description="메타데이터")
    context_data: Optional[str] = Field(None, description="임베딩 제외 컨텍스트 데이터 (SQL, 스키마 등 Agent 참조용)")

    model_config = {
        "json_schema_extra": {
            "example": {
                "title": "2024년 인사규정 (수정)",
                "content": "제1조 목적... (수정된 내용)"
            }
        }
    }


class DocumentUpdateResponse(BaseModel):
    """문서 수정 응답"""
    success: bool
    message: str
    doc_id: int
    title: str
    content_length: int
    embedding_invalidated: bool
    needs_reindex: bool


class DocumentDeleteResponse(BaseModel):
    """문서 삭제 응답"""
    success: bool
    message: str
    deleted_count: int


class BulkDelete:
    """일괄 삭제 관련 스키마 그룹 (Inner Class 패턴)"""

    class Request(BaseModel):
        """일괄 삭제 요청"""
        doc_ids: List[int] = Field(..., min_length=1, description="삭제할 문서 ID 목록")

        model_config = {
            "json_schema_extra": {
                "example": {
                    "doc_ids": [1, 2, 3]
                }
            }
        }

    class Response(BaseModel):
        """일괄 삭제 응답"""
        success: bool
        message: str
        total_requested: int
        total_deleted: int
        failed_ids: List[int] = Field(default_factory=list)


# ===================================
# 청킹(Chunking) 관련
# ===================================

class Chunking:
    """청킹 관련 스키마 그룹 (Inner Class 패턴)"""

    class PreviewRequest(BaseModel):
        """청킹 미리보기 요청"""
        content: str = Field(..., min_length=1, description="문서 내용")
        chunk_size: int = Field(default=1000, ge=100, le=5000, description="청크 크기")
        chunk_overlap: int = Field(default=100, ge=0, le=500, description="청크 간 중복")

    class PreviewItem(BaseModel):
        """청킹 미리보기 항목"""
        index: int
        length: int
        preview: str

    class PreviewResponse(BaseModel):
        """청킹 미리보기 응답"""
        original_length: int
        total_chunks: int
        chunks: List["Chunking.PreviewItem"]

    class ExecuteRequest(BaseModel):
        """청킹 실행 요청"""
        doc_ids: List[int] = Field(..., min_length=1, description="청킹할 문서 ID 목록")
        chunk_size: int = Field(default=1000, ge=100, le=5000, description="청크 크기 (문자 수)")
        chunk_overlap: int = Field(default=100, ge=0, le=500, description="청크 간 중복 (문자 수)")
        delete_original: bool = Field(default=False, description="원본 문서 삭제 여부")

        model_config = {
            "json_schema_extra": {
                "example": {
                    "doc_ids": [1, 2, 3],
                    "chunk_size": 1000,
                    "chunk_overlap": 100,
                    "delete_original": False
                }
            }
        }

    class ExecuteResultItem(BaseModel):
        """청킹 실행 결과 항목"""
        original_doc_id: int
        original_title: str
        success: bool
        error: Optional[str] = None
        parent_id: Optional[int] = None
        chunk_ids: List[int] = Field(default_factory=list)
        total_chunks: int = 0
        original_chars: int = 0

    class ExecuteResponse(BaseModel):
        """청킹 실행 응답"""
        success: bool
        message: str
        total_requested: int
        total_success: int
        total_failed: int
        results: List["Chunking.ExecuteResultItem"]


# Pydantic v2에서 forward reference 해결
Chunking.PreviewResponse.model_rebuild()
Chunking.ExecuteResponse.model_rebuild()
